cleanlab_studio/internal/util.py

```diff
@@ -50,6 +63,124 @@ def init_dataset_source(
         raise ValueError("Invalid dataset source provided")
 
 
+def apply_corrections_snowpark_df(
+    dataset: Any,
+    cl_cols: Any,
+    id_col: str,
+    label_column: str,
+    keep_excluded: bool,
+) -> Any:
+    from snowflake.snowpark.functions import (
+        when,
+        col,
+        is_null,
+        monotonically_increasing_id,
+    )
+
+    label_col = quote(label_column)
+    id_col = quote(id_col)
+    action_col = quote("action")
+    corrected_label_col = quote("corrected_label")
+    cleanlab_final_label_col = quote("__cleanlab_final_label")
+
+    corrected_ds = dataset
+    session = dataset.session
+
+    cl_cols = session.create_dataframe(cl_cols)
+
+    if id_col not in corrected_ds.columns:
+        corrected_ds = corrected_ds.withColumn(id_col, monotonically_increasing_id())
+
+    both = cl_cols.select([id_col, action_col, corrected_label_col]).join(
+        corrected_ds.select([id_col, label_col]),
+        on=id_col,
+        how="left",
+    )
+
+    final = both.withColumn(
+        cleanlab_final_label_col,
+        when(is_null(corrected_label_col), col(label_col)).otherwise(col(corrected_label_col)),
+    )
+
+    new_labels = final.select([id_col, action_col, cleanlab_final_label_col]).withColumnRenamed(
+        cleanlab_final_label_col, label_col
+    )
+
+    corrected_ds = corrected_ds.drop(label_col).join(new_labels, on=id_col, how="left")
+    corrected_ds = (
+        corrected_ds.where((col(action_col) != "exclude") | is_null(col(action_col)))
+        if not keep_excluded
+        else corrected_ds
+    ).drop(action_col)
```

Comment:
I think we can clean this up a bit by just doing:

- select id_col, action_col, corrected_label_col from cl_cols
- left join with original dataset on id_col
- temporarily add final label col
- drop original label col and corrected_label_col
- rename final label col to original label col name
- filter out excluded rows
- drop action col

This removes the need to do the multiple joins.